core/src/main/java/com/linecorp/armeria/internal/common/ClientHttp2ObjectEncoder.java

```diff
@@ -0,0 +1,105 @@
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.stream.ClosedStreamException;
+import com.linecorp.armeria.internal.HttpHeaderUtil;
+
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http2.Http2Connection;
+import io.netty.handler.codec.http2.Http2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2Headers;
+
+public final class ClientHttp2ObjectEncoder extends Http2ObjectEncoder {
+    private final SessionProtocol protocol;
+
+    public ClientHttp2ObjectEncoder(ChannelHandlerContext ctx, Http2ConnectionEncoder encoder,
+                                       SessionProtocol protocol) {
+        super(ctx, encoder);
+        this.protocol = requireNonNull(protocol, "protocol");
+    }
+
+    @Override
+    protected ChannelFuture doWriteHeaders(int id, int streamId, HttpHeaders headers, boolean endStream,
+                                           HttpHeaders additionalHeaders, HttpHeaders additionalTrailers) {
+        final Http2Connection conn = encoder().connection();
+        final boolean isTrailer = !headers.contains(HttpHeaderNames.METHOD);
+        final Http2Headers convertedHeaders;
+
+        if (isStreamPresentAndWritable(streamId)) {
+            if (!isTrailer) {
+                convertedHeaders = convertHeaders(headers, additionalHeaders, additionalTrailers);
+            } else {
+                convertedHeaders = convertTrailers(headers, additionalTrailers);
+            }
+            // Writing to an existing stream.
+            return encoder().writeHeaders(ctx(), streamId, convertedHeaders, 0, endStream,
+                                          ctx().newPromise());
+        }
+
+        if (conn.local().mayHaveCreatedStream(streamId)) {
+            // Stream has been closed.
+            return newFailedFuture(ClosedStreamException.get());
+        }
+
+        if (!isTrailer) {
+            convertedHeaders = convertHeaders(headers, additionalHeaders, additionalTrailers);
+        } else {
+            convertedHeaders = convertTrailers(headers, additionalTrailers);
+        }
+
+        // Client starts a new stream.
+        return encoder().writeHeaders(ctx(), streamId, convertedHeaders, 0, endStream,
+                                      ctx().newPromise());
+    }
+
+    private Http2Headers convertHeaders(
+            HttpHeaders inputHeaders, HttpHeaders additionalHeaders, HttpHeaders additionalTrailers) {
+        final Http2Headers outputHeaders =
+                ArmeriaHttpUtil.toNettyHttp2Client(inputHeaders, additionalHeaders, additionalTrailers, false);
+
+        if (!outputHeaders.contains(HttpHeaderNames.USER_AGENT)) {
+            outputHeaders.add(HttpHeaderNames.USER_AGENT, HttpHeaderUtil.USER_AGENT.toString());
+        }
+
+        if (!outputHeaders.contains(HttpHeaderNames.SCHEME)) {
+            outputHeaders.scheme(protocol.isTls() ? SessionProtocol.HTTPS.uriText()
```

Comment:
How about using `add()` instead of shortcut methods for better performance?

Reply:
Fixed it :+1: